### PR TITLE
Add Composer

### PR DESCRIPTION
**Pull request by Hypolite**
After much consideration, Friendica just reached the state of the art of PHP development in 2012 by adding support for Composer.

For now, it doesn't change anything to anyone workflow, whether they are developer, cutting-edge users or stable users. The Friendica repository still bundles all the library dependencies and a simple `git pull` still get you the whole codebase required to run Friendica.

Long term goal is to move all the current libraries in `library` to dependencies in `composer.json` that get downloaded in `vendor`. If you participate to this effort, just run `util/composer.phar update` from Friendica install directory and commit all the changes.

This PR also sets the Friendica namespace folder to `src` so we can visibly see how much of the `include` directory has been converted to proper classes. If you care about performance, you can run `util/composer update` after each converted class, but it's not mandatory.

Down the road, once `library` and `vendor` will have been cleared, we will be able to use `util/composer.phar archive` to generate a self-contained archive including all the required dependencies, for releases for example. This process could be automated to provide weekly builds for example. An install/update script will be published including running a `util/composer.phar install` command to properly set up Friendica's dependancies and autoload mechanism.

Even further, we will be able to distribute Friendica directly through Composer's main repository, Packagist. This would require us to list all the PHP requirements so that a warning can be shown when someone tries to install Friendica through Composer.

### CODE REVIEW

File: boot.php [hunk 1]
```diff
@@ -17,7 +17,7 @@
  * easily as email does today.
  */
 
-require_once('include/autoloader.php');
+require_once(__DIR__ . DIRECTORY_SEPARATOR. 'vendor' . DIRECTORY_SEPARATOR . 'autoload.php');
```
Comment (annando) [inline boot.php]: Do you like to remove the brackets from all ```require_once``` in this file?

Comment (Hypolite) [inline boot.php]: Yes, but @Quix0r is doing it in his massive PR, and I don't want to trigger more conflicts than what this will already do.

**Comment by Hypolite** [comment]
I'm fine with documentation also. However, I'm not sure where to put it, and I'd like to do it in a future PR so that the current issues waiting on Composer can be processed quicker.

**Comment by annando** [comment]
Documentation is mostly observed by @tobiasd - he should decide.

**Comment by Hypolite** [comment]
My main question is whether it should be in the Wiki or in Friendica built-in docs. Either way I can start writing something.

**Comment by annando** [comment]
The wiki is only some backup thing, since it is some centralized thing. Just add a few lines in ```Developers-Intro.md```.

**Comment by Hypolite** [comment]
Good, I'm doing this right now.

**Comment by tobiasd** [comment]
On Sat, 01 Apr 2017 09:56:20 -0700
Hypolite Petovan <notifications@github.com> wrote:

> My main question is whether it should be in the Wiki or in Friendica
> built-in docs

In the build in docs, so everybody has it on their own fingertips. These docs are also liked from the wiki as "The Documentation".

File: boot.php (hunk 1 above)
Comment (Quix0r) [inline boot.php]: I like the idea of using `DIRECTORY_SEPARATOR`. It is way more cross-platform compatible.

**Comment by Quix0r** [comment]
There is a way of having both documentations at the same time. Maybe obvious, use markdown code. :-)

**Comment by Hypolite** [comment]
Documentation added.

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+		"autoloader-suffix": "Friendica",
+		"optimize-autoloader": true,
+		"preferred-install": "dist"
+	},
+	"archive": {
+		"exclude": [
+			"log", "cache", "/photo", "/proxy"
+		]
+	}
+}
```
Comment (Quix0r) [inline composer.json]: New-line missing at end of file.

**Comment by annando** [comment]
@Hypolite Since you had ask me and @fabrixxm for a review: Do you want to wait for his expertise or may I accept your PR as well?

File: doc/Home.md [hunk 3]
```diff
@@ -47,6 +47,7 @@ Friendica Documentation and Resources
 * [Protocol Documentation](help/Protocol)
 * [Database schema documantation](help/database)
 * [Class Autoloading](help/autoloader)
+* [Using Composer](help/Composer)
```
Comment (tobiasd) [inline doc/Home.md]: Can you add this link to `doc/de/Home.md` as well and add a `((EN)` behind it as indicator for the text being available only in English at the moment.

Comment (Hypolite) [inline doc/Home.md]: Done, thanks for the reminder.

**Comment by Hypolite** [comment]
Just pulled `develop`, no issue. 🎉

**Comment by annando** [comment]
I had some php errors during the pulling - but this was expected.

**Comment by Hypolite** [comment]
Filing into install/update script: put the instance in maintenance mode before pulling/composing.

**Comment by AlfredSK** [comment]
git pull was successful. I've prepared my update script for composer already but commented out the line 
```util/composer.phar install```
Question is from when do I have to uncomment the line? How do I know the right time to switch to composer?

**Comment by Hypolite** [comment]
@AlfredSK You *can* already use it without a problem, just you currently don't *need* to, since we are shipping the `vendor` folder already populated and ready-to-use. I'd suggest you don't comment out the command so that when we eventually clear the vendor folder you'll be already set up.

**Comment by Quix0r** [comment]
No, `vendor` should not be included and ignored in `.gitignore` as it is done by many projects and `composer install` will prepare `./vendor/` for you.

**Comment by Quix0r** [comment]
I have committed id 459060e in my `develop` branch. You can cherry-pick it:

````
git remote add quix0r git@github.com:Quix0r/friendica.git
git cherry-pick 459060e
````

This way you don't "steal" "my" commit (the author remains then me and not becomes you).

**Comment by annando** [comment]
@Quix0r we will do so in a next step.

**Comment by Quix0r** [comment]
@annando just cherry-pick my commit. :-)

**Comment by Quix0r** [comment]
I forgot to delete them before I committed + ignore them. Now included with new commit id (force-pushed).

**Comment by annando** [comment]
This will be done some time later. We will eventually do this when the installation and the update for regular users has been improved.

**Comment by Quix0r** [comment]
Updated my installation to composer-based. Here is my (still) simple [deploy.sh](https://f.haeder.net/deploy.sh) script.

**Comment by AlfredSK** [comment]
@Hypolite Okay. Thanks. I tried the composer command and got a warning about not to run it as root. See https://get composer.org/root
The suggestion is to use ```composer install --no-plugins --no-scripts``` for security reasons. Is this possible or do I break something if I use these options?
Alternatively I could run the composer command as 'www-data'. But then it complains about not being able to create the composer cache directory in /root/.composer
Ideas?

**Comment by Quix0r** [comment]
A bit OT: I propose to have a "staging" branch around which will be updated when the code is considered as "stable enough". Then people like @AlfredSK can switch to it (`git checkout some-branch`), else with next push maybe something is broken on `develop`.

Then they are still "bleeding-edge" but not to close to development.

**Comment by Quix0r** [comment]
@AlfredSK why **root**?

**Comment by annando** [comment]
@Quix0r we need people like @AlfredSK who are on the bleeding edge since we need testers.

Please don't let us start the discussion again. We will do the recommend way as soon as we find a good way so that users, testers and developers are lucky about the situation.

**Comment by Hypolite** [comment]
@AlfredSK You can use www-data home dir with `sudo -H`, or if www-data doesn't have a home directory, you can use `COMPOSER_HOME=/tmp/composer util/composer.phar install` to avoid such issue.

**Comment by annando** [comment]
We should collect these problems, so that we can add them to the documentation.

**Comment by Hypolite** [comment]
I'm not sure about that, this sounds like @AlfredSK is doing something quite wrong on his system by wandering around as root in the first place. I'm afraid this would fall into a way more general "Server Admin FAQ" that I don't think we should be editing.

**Comment by Hypolite** [comment]
Gotcha, I'll update the Composer Help page with a starting FAQ.

**Comment by AlfredSK** [comment]
Thanks @Hypolite for the hint about COMPOSER_HOME. In combination with sudo -u www-data it works without any warnings now. 🎉 🍺

**Comment by Hypolite** [comment]
@AlfredSK What about `sudo -H -u www-data`, did it show warnings as well? I'm writing a Composer FAQ and I would like to choose the simplest solution.

**Comment by AlfredSK** [comment]
```sudo -H -u www-data util/composer.phar install    Cannot create cache directory /var/www/.composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Cannot create cache directory /var/www/.composer/cache/files/, or directory is not writable. Proceeding without cache
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Generating optimized autoload files```

**Comment by AlfredSK** [comment]
Btw. I set COMPOSER_HOME to /var/tmp/composer so the directory survives reboots.

**Comment by Hypolite** [comment]
Ok, makes sense, `/var/www` is `www-data` home, but it's probably owned by `root`.

**Pull request merged**